### ISSUE TITLE
[SC3028] In a `if [ -n BASH_SOURCE ]` block don't warn of BASH_SOURCE being undefined

### ISSUE THREAD

**Issue by SamuelMarks** [koalaman/shellcheck#3111]
#### For bugs
- Rule Id (if any, e.g. SC1000): SC3028
- My shellcheck version (`shellcheck --version` or "online"): 0.10.0
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [x] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### For new checks and feature suggestions
- [x] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/sh

if [ -n "${BASH_VERSION}" ]; then
  # shellcheck disable=SC3028 disable=SC3054
  this_file="${BASH_SOURCE[0]}"
elif [ -n "${ZSH_VERSION}" ]; then
  # shellcheck disable=SC2296
  this_file="${(%):-%x}"
else
  this_file="${0}"
fi
```

#### Here's what shellcheck currently says:
When I remove the comments it says:
```
In POSIX sh, BASH_SOURCE is undefined.
See SC3028.
In POSIX sh, array references are undefined.
See SC3054.
```

#### Here's what I wanted or expected to see:

No error

**Comment by SamuelMarks** [comment]
@brother This is my actual project with relevant line highlighted https://github.com/SamuelMarks/deploy-sh/blob/7b03f64/install.sh#L6

As you can see I've had to include a large number of `# shellcheck disable` 🤦

**Comment by brother** [comment]
That doesn't answer my question.
The title of the issue and the code example differs. It's hard to follow.
![bild](https://github.com/user-attachments/assets/163910ce-9f02-4b59-a0fc-5332599cfc3f)

I don't really see how shellcheck could behave different. You asked it to use sh/posix mode and then check one variable and want another one to be included too?

**Comment by SamuelMarks** [comment]
@brother Removed the ambiguity from the issue title.

Basically if I have an `if` check to see if I'm in Bash then I want to be able to use Bash features without using `# shellcheck disable` everywhere.
